contrib/insights/src/column_analyzer.cxx

```diff
@@ -0,0 +1,120 @@
+#include <functional>
+#include <string>
+#include <cmath>
+#include <limits>
+#include <map>
+#include <utility>
+
+
+using namespace tbb;
+
+namespace insights {
+namespace detail {
+
+
+void column_analyzer::operator()(const blocked_range<size_t>& r) const {
+    for (size_t feature_idx = r.begin(); feature_idx != r.end(); ++feature_idx) {
+        auto t = (*current_example)->features().at(feature_idx);
+        column_analysis & feature_statistics = columns->at(feature_idx);
+        
+        auto dt = static_cast<mlio::dense_tensor*>(t.get());
+        auto cells = dt->data().as<std::string>();
+
+        auto should_capture = capture_columns->count(feature_idx) > 0;
+
+        // Used to update the mean.
+        // We do this here to aggregate several entries together and avoid
+        // potential numerical problems updating the mean a single entry at a time. 
+        double numeric_column_sum = 0.0;
+        double numeric_column_count = 0.0;
+
+        for (std::string as_string : cells) {
+            feature_statistics.rows_seen += 1;
+            feature_statistics.example_value = as_string;
+            
+            if (should_capture) {
+                if (feature_statistics.string_captured_unique_values.size() < max_capture_count) {
+                    feature_statistics.string_captured_unique_values.insert(as_string);
+                } else if (feature_statistics.string_captured_unique_values.count(as_string) == 0) {
+                    // If the value isn't present but we're not adding it because we're at a limit
+                    // then we should flag that we have overflowed.
+                    feature_statistics.string_captured_unique_values_overflowed = true;
+                }
+            }
+
+            // String analyzers.
+            feature_statistics.string_cardinality_estimator.add(as_string);
+            
+            feature_statistics.string_min_length = 
+                std::min(feature_statistics.string_min_length, as_string.size());
+            
+            feature_statistics.string_max_length = 
+                std::max(feature_statistics.string_max_length, as_string.size());
+
+            if (as_string.size() == 0) {
+                feature_statistics.string_empty_count += 1;
+            } else {
+                feature_statistics.string_min_length_not_empty = 
+                    std::min(feature_statistics.string_min_length_not_empty, as_string.length());
+            }
+
+            if (mlio::only_whitespace(as_string)) {
+                feature_statistics.string_only_whitespace_count += 1;
+            }
+
+            auto lowercase_string = as_string;
+            std::transform(as_string.begin(), as_string.end(), lowercase_string.begin(), ::tolower);
+            if (mlio::matches(lowercase_string, null_like_values)) {
+                feature_statistics.string_null_like_count += 1;
+            }
+
+            // Numeric analyzers
+            double as_float;
```

Comment:
What about integer fields? Should we not have a distinction between integers and floating-point numbers for statistics?

Reply:
Good point. We pulled them out because they weren't originally necessary but I don't think there's any harm in doing int-based analysis. I'll add it to a second iteration/pass through.